### PR TITLE
[next] UI Components

### PR DESCRIPTION
**Pull request by okbel**
### What does this PR do?
- Adds UI components needed for Stage #2 of V5 - Mostly for Auth
- Extends `Icon` Component adding extra small property - `xs`
- Extra page to show Form Examples

### Components
- `ValidationMessage`
- `TextField`
- `InputLabel`
- `CallOut`

### How to test them?
You can go to *Forms* in the docs to see them in action.

### CODE REVIEW

File: src/core/client/ui/components/Icon/Icon.mdx [hunk 1]
```diff
@@ -14,6 +14,7 @@ Checkout available icons https://material.io/tools/icons/
 
 ## Basic usage
 <Playground>
+  <Icon size="xsm">face</Icon>
```
Comment (alinhle) [inline src/core/client/ui/components/Icon/Icon.mdx]: In the Permalink PR https://github.com/coralproject/talk/pull/1741/files#diff-d17b1e6852ac82c4830bbb5ff488c5edR40, the icon size `sm` has a `14px`, so we don't need to add a new size here.

File: src/core/client/ui/components/Icon/Icon.css [hunk 2]
```diff
@@ -36,6 +36,10 @@
   -moz-osx-font-smoothing: grayscale;
 }
 
+.xs {
+  font-size: 14px;
+  width: 14px;
+}
```
Comment (alinhle) [inline src/core/client/ui/components/Icon/Icon.css]: In the Permalink PR https://github.com/coralproject/talk/pull/1741/files#diff-d17b1e6852ac82c4830bbb5ff488c5edR40, the icon size sm has a 14px, so we don't need to add a new size here.

File: src/core/client/ui/components/Icon/Icon.tsx [hunk 3]
```diff
@@ -13,7 +13,7 @@ interface InnerProps extends HTMLAttributes<HTMLSpanElement> {
    */
   classes: typeof styles;
 
-  size?: "sm" | "md" | "lg" | "xl";
+  size?: "xs" | "sm" | "md" | "lg" | "xl";
```
Comment (alinhle) [inline src/core/client/ui/components/Icon/Icon.tsx]: In the Permalink PR https://github.com/coralproject/talk/pull/1741/files#diff-d17b1e6852ac82c4830bbb5ff488c5edR40, the icon size sm has a 14px, so we don't need to add a new size here.

File: src/core/client/ui/components/InputLabel/InputLabel.tsx [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+  classes: typeof styles;
+}
+
+const InputLabelProps: StatelessComponent<InputLabelProps> = props => {
+  const { className, children, classes, ...rest } = props;
+
+  const rootClassName = cn(classes.root, className);
+
+  return (
+    <Typography className={rootClassName} {...rest}>
```
Comment (alinhle) [inline src/core/client/ui/components/InputLabel/InputLabel.tsx]: How about adding `inputLabel` as a variant to `Typography` and use it?

File: src/core/client/ui/components/TextField/TextField.tsx [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+   */
+  classes: typeof styles;
+  /**
+   * Color of the TextField
+   */
+  color?: "regular" | "error";
+  /*
+  * If set renders a full width button
+  */
+  fullWidth?: boolean;
```
Comment (alinhle) [inline src/core/client/ui/components/TextField/TextField.tsx:31]: Here we need the `placeholder` support

File: src/core/client/ui/components/ValidationMessage/ValidationMessage.css [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+  composes: validationMessage from "talk-ui/shared/typography.css";
+  position: relative;
+  display: inline-flex;
+  justify-content: flex-start;
+  align-items: center;
+  padding: 7px 10px;
+  box-sizing: border-box;
+}
+
+.colorRegular {
```
Comment (alinhle) [inline src/core/client/ui/components/ValidationMessage/ValidationMessage.css]: This styling looks a bit odd

File: src/core/client/ui/components/ValidationMessage/ValidationMessage.tsx [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+      [classes.colorError]: color === "error",
+      [classes.fullWidth]: fullWidth,
+    },
+    className
+  );
+
+  return (
+    <div className={rootClassName} {...rest}>
+      {color === "error" && (
+        <Icon size="xs" className={classes.icon}>
```
Comment (alinhle) [inline src/core/client/ui/components/ValidationMessage/ValidationMessage.tsx]: We can use `size="sm"` here, see above.

File: src/core/client/ui/components/CallOut/CallOut.css [hunk 8]
```diff
@@ -0,0 +1,35 @@
+.root {
+  composes: bodyCopy from "talk-ui/shared/typography.css";
+  position: relative;
+  display: inline-flex;
+  justify-content: center;
+  align-items: center;
+  padding: 12px;
```
Comment (alinhle) [inline src/core/client/ui/components/CallOut/CallOut.css]: Most of the time we use a multiple of `spacing-unit`. This should be `padding: var(--spacing-unit)`.

File: src/core/client/ui/components/TextField/TextField.css [hunk 9]
```diff
@@ -0,0 +1,27 @@
+.root {
+  composes: textField from "talk-ui/shared/typography.css";
+  position: relative;
+  display: inline-flex;
+  padding: 7px 10px;
```
Comment (alinhle) [inline src/core/client/ui/components/TextField/TextField.css]: What I see from the design, this should be `padding: calc(0.5 * var(--spacing-unit));`

File: src/core/client/ui/components/TextField/TextField.css [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+  box-sizing: border-box;
+  border-radius: 1px;
```
Comment (alinhle) [inline src/core/client/ui/components/TextField/TextField.css]: We have the border radius also as a theme variable :-) `border-radius: var(--round-corners);`

File: src/core/client/ui/components/TextField/TextField.css [hunk 11]
```diff
@@ -0,0 +1,27 @@
+.root {
+  composes: textField from "talk-ui/shared/typography.css";
+  position: relative;
+  display: inline-flex;
```
Comment (alinhle) [inline src/core/client/ui/components/TextField/TextField.css]: Why is it set to `inline-flex`?

File: src/core/client/ui/components/TextField/TextField.css [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+.colorError {
+  background-color: var(--palette-common-white);
+  border-color: var(--palette-error-main);
+  border: 2px solid #FA4643;
+}
+
+.fullWidth {
+  display: block;
+  width: 100%;
+  box-sizing: border-box;
```
Comment (alinhle) [inline src/core/client/ui/components/TextField/TextField.css]: This has already been set at `.root`.

File: src/core/client/ui/components/ValidationMessage/ValidationMessage.css [hunk 13]
```diff
@@ -0,0 +1,35 @@
+.root {
+  composes: validationMessage from "talk-ui/shared/typography.css";
+  position: relative;
+  display: inline-flex;
+  justify-content: flex-start;
+  align-items: center;
+  padding: 7px 10px;
```
Comment (alinhle) [inline src/core/client/ui/components/ValidationMessage/ValidationMessage.css]: `padding: calc(0.5 * var(--spacing-unit)) var(--spacing-unit)`

File: src/core/client/ui/components/ValidationMessage/ValidationMessage.css [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+  border: none;
+  color: var(--palette-common-black);
+  padding-left: 40px;
+}
+
+.colorError {
+  background-color: var(--palette-error-light);
+  border-radius: 1px;
+  border-color: var(--palette-error-darkest);
+  border-left-width: 8px;
```
Comment (alinhle) [inline src/core/client/ui/components/ValidationMessage/ValidationMessage.css]: Hm weird spacing here, will need to ask @IAmSamHankins if this shouldn't be `0.5 * spacing-unit`

File: src/core/client/ui/components/ValidationMessage/ValidationMessage.css [hunk 15]
```diff
@@ -0,0 +1,35 @@
...
+  border-left-style: solid;
+  color: var(--palette-common-white);
+}
+
+.fullWidth {
+  display: block;
+  width: 100%;
+  box-sizing: border-box;
```
Comment (alinhle) [inline src/core/client/ui/components/ValidationMessage/ValidationMessage.css]: Already set in `.root`

File: src/core/client/ui/components/ValidationMessage/ValidationMessage.css [hunk 16]
```diff
@@ -0,0 +1,35 @@
...
+  background-color: var(--palette-error-light);
+  border-radius: 1px;
+  border-color: var(--palette-error-darkest);
+  border-left-width: 8px;
+  border-left-style: solid;
+  color: var(--palette-common-white);
+}
+
+.fullWidth {
+  display: block;
```
Comment (alinhle) [inline src/core/client/ui/components/ValidationMessage/ValidationMessage.css]: This should be `display: flex`

File: src/core/client/ui/components/ValidationMessage/ValidationMessage.css [hunk 17]
```diff
@@ -0,0 +1,35 @@
...
+  width: 100%;
+  box-sizing: border-box;
+}
+
+.icon {
+  margin-right: 8px;
```
Comment (alinhle) [inline src/core/client/ui/components/ValidationMessage/ValidationMessage.css]: Hm weird spacing here, will need to ask @IAmSamHankins if this shouldn't be 1 * spacing-unit

**Review by alinhle — CHANGES_REQUESTED** [review]
Tiny comments, overall it looks good!

I saw you upgraded every dependency, for some reason this breaks docz and jest.
In general I would avoid upgrading dependencies unless they are necessary to fullfill the story, and then I would only upgrade related dependencies. Lets make a separate PR if we want to make an major upgrade where me make sure that everything works especially as some packages we are using are bleeding edge.

In order to restore docz and jest I have found these 2 steps will resolve it:

  - [ ] Merge or wait for https://github.com/coralproject/talk/pull/1782
  - [ ] Revert upgrading dependencies

File: src/core/client/ui/components/CallOut/CallOut.css [hunk 18]
```diff
@@ -0,0 +1,35 @@
...
+.colorError {
+  background-color: var(--palette-error-lightest);
+  border-color: var(--palette-error-light);
+  color: var(--palette-text-primary);
+}
+
+.fullWidth {
+  display: block;
+  width: 100%;
+  box-sizing: border-box;
```
Comment (alinhle) [inline src/core/client/ui/components/CallOut/CallOut.css]: This is already set in `.root`

File: src/core/client/ui/components/CallOut/CallOut.css [hunk 19]
```diff
@@ -0,0 +1,35 @@
...
+}
+
+.colorError {
+  background-color: var(--palette-error-lightest);
+  border-color: var(--palette-error-light);
+  color: var(--palette-text-primary);
+}
+
+.fullWidth {
+  display: block;
```
Comment (alinhle) [inline src/core/client/ui/components/CallOut/CallOut.css]: This should be `display: flex`.

File: src/core/client/ui/components/ValidationMessage/ValidationMessage.css (hunk 15 above)
Comment (alinhle) [reply src/core/client/ui/components/ValidationMessage/ValidationMessage.css]: You probably took that from `Button`, I fixed the `Button.css` in the permalink PR.

File: src/core/client/ui/components/Icon/Icon.mdx (hunk 1 above)
Comment (okbel) [reply src/core/client/ui/components/Icon/Icon.mdx]: Then I will replace it for `sm` and wait for Permalink to be merge so I can check and match styles.

File: src/core/client/ui/components/Icon/Icon.css (hunk 2 above)
Comment (okbel) [reply src/core/client/ui/components/Icon/Icon.css]: 👍

File: src/core/client/ui/components/InputLabel/InputLabel.tsx (hunk 4 above)
Comment (okbel) [reply src/core/client/ui/components/InputLabel/InputLabel.tsx]: Hmmm, yeah. This is just a wrapper for Typography after all. But I like the API. Will think about it :p

File: src/core/client/ui/components/TextField/TextField.tsx (hunk 5 above)
Comment (okbel) [reply src/core/client/ui/components/TextField/TextField.tsx:31]: 👍

File: src/core/client/ui/components/ValidationMessage/ValidationMessage.css (hunk 13 above)
Comment (okbel) [reply src/core/client/ui/components/ValidationMessage/ValidationMessage.css]: Wow. That line is ugly. Maybe we can add some extra func here. Adding the line 👍 .

File: src/core/client/ui/components/ValidationMessage/ValidationMessage.css (hunk 14 above)
Comment (okbel) [reply src/core/client/ui/components/ValidationMessage/ValidationMessage.css]: I used exactly wha Figma showed.

File: src/core/client/ui/components/TextField/TextField.css (hunk 11 above)
Comment (okbel) [reply src/core/client/ui/components/TextField/TextField.css]: Why not? Do you think it should be `block`?

**Comment by okbel** [comment]
@cvle Docz PR was merged and reversing the upgrade didn't quite resolve the issue. 
On the other hand, `jest` was not upgraded and it's triggering this error.

```
  ● Test suite failed to run

    Some errors occurred while attempting to read from /Users/bel/Develop/talk/config/jest/tsconfig.jest.json: error TS5058: The specified path does not exist: '/Users/bel/Develop/talk/config/jest/tsconfig.json'.
    error TS18003: No inputs were found in config file 'tsconfig.json'. Specified 'include' paths were '["**/*"]' and 'exclude' paths were '[]'.

      at readCompilerOptions (node_modules/ts-jest/dist/utils/get-ts-config.js:56:15)
      at Object.getTSConfig_local (node_modules/ts-jest/dist/utils/get-ts-config.js:31:18)
      at Object.memoized [as default] (node_modules/lodash/lodash.js:10551:27)
      at Object.getCacheKey (node_modules/ts-jest/dist/utils/get-cache-key.js:17:32)
```

**Comment by okbel** [comment]
We can discuss this further in Arch Meeting but I think we can progressively keep our dependencies up to date. We can schedule that every two weeks (or so) we run an upgrade on Talk and check that everything still works as expected. Falling behind with versions is an issue that I think, we should avoid.

**Comment by okbel** [comment]
Changes added. Fully working. Ready to be checked @cvle

File: src/core/client/ui/components/TextField/TextField.css [hunk 20]
```diff
@@ -1,8 +1,25 @@
...
+  position: relative;
+  display: block;
+  padding: calc(0.5 * var(--spacing-unit));
   box-sizing: border-box;
   border-radius: var(--round-corners);
-  padding: calc(0.5 * var(--spacing-unit));
 }
+
+.colorRegular {
+  backgroun-color: var(--palette-common-white);
```
Comment (alinhle) [inline src/core/client/ui/components/TextField/TextField.css]: Here is a little typo :-)

File: src/core/client/ui/components/TextField/TextField.mdx [hunk 21]
```diff
@@ -4,14 +4,18 @@ menu: UI Kit
...
-`TextField`
-
-## Basic usage
+## Basic Use
 <Playground>
-  <TextField value="Hallo Talk" />
+  <Flex itemGutter direction="column">
+    <TextField defaultValue="This is an input field" />
+    <TextField color="error" defaultValue="A TextField with an error" />
+    <TextField color="error" defaultValue="A TextField with an error" fullWidth/>
```
Comment (alinhle) [inline src/core/client/ui/components/TextField/TextField.mdx]: Let's add a placeholder example here

File: src/core/client/ui/components/TextField/TextField.css [hunk 22]
```diff
@@ -1,8 +1,25 @@
...
+.colorError {
+  background-color: var(--palette-common-white);
+  border-color: var(--palette-error-main);
+  border: 2px solid #FA4643;
+}
+
+.fullWidth {
+  width: 100%;
+}
+
```
Comment (cvle) [inline src/core/client/ui/components/TextField/TextField.css]: We should add the styling for placeholder here, see figma file: https://www.figma.com/file/NdlkTfgQNPSglYvYUoVVw6EH/Coral-Styles?node-id=60%3A434

File: src/core/client/ui/components/TextField/TextField.css [hunk 23]
```diff
@@ -1,8 +1,33 @@
...
+::placeholder {
+  font-family: var(--font-family);
+  font-style: normal;
+  font-weight: normal;
+  line-height: normal;
+  font-size: calc(16rem / var(--rem-base));
+  color: var(--palette-grey-lighter);
```
Comment (alinhle) [inline src/core/client/ui/components/TextField/TextField.css]: Let's move this to typography.css, so we match with the figma typographies

File: src/core/client/ui/shared/typography.css [hunk 24]
```diff
@@ -127,11 +127,29 @@
...
 }
 
-.textField {
+.validationMessage {
   color: var(--palette-common-black);
-  font-family: "Source Sans Pro";
+  font-family: var(--font-family);
+  font-weight: var(--font-weight-medium);
+  font-size: calc(16rem / var(--rem-base));
+  line-height: normal;
```
Comment (alinhle) [inline src/core/client/ui/shared/typography.css]: Use line height from figma designs, should be `calc(16rem / var(--rem-base));`

File: src/core/client/ui/shared/typography.css [hunk 25]
```diff
@@ -127,11 +127,29 @@
   letter-spacing: calc(0.2em / 16);
 }
 
-.textField {
+.validationMessage {
```
Comment (alinhle) [inline src/core/client/ui/shared/typography.css]: Just synced with @IAmSamHankins, and we want to keep the name `alertMessage` as defined in the designs.

Comment (alinhle) [reply src/core/client/ui/shared/typography.css]: only the css name, not the component name.

File: src/core/client/ui/shared/typography.css (hunk 24 above)
Comment (okbel) [reply src/core/client/ui/shared/typography.css]: Why is it different from normal? - And Figma says `normal`. Do you have another config? I'm checking the CSS tab

File: src/core/client/ui/components/TextField/TextField.css (hunk 23 above)
Comment (okbel) [reply src/core/client/ui/components/TextField/TextField.css]: 👍

Comment (okbel) [reply src/core/client/ui/components/TextField/TextField.css]: I think it's this decision is okay but `[Error] Error: composition is only allowed when selector is single :local class name not in "::placeholder", "::placeholder" is weird` I'm getting that error.

**Review by alinhle — APPROVED** [review]

**Pull request merged**